components/codec/number.rs

```diff
@@ -0,0 +1,2096 @@
+    ///
+    /// # Panics
+    ///
+    /// Panics when `buf.len() < 8`.
+    #[inline]
+    pub fn decode_i64_le(buf: &[u8]) -> i64 {
+        LittleEndian::read_i64(buf)
+    }
+
+    /// Encodes a 64 bit float number `v` to `buf` in little endian,
+    /// which is not memory-comparable.
+    ///
+    /// # Panics
+    ///
+    /// Panics when `buf.len() < 8`.
+    #[inline]
+    pub fn encode_f64_le(buf: &mut [u8], v: f64) {
+        LittleEndian::write_f64(buf, v)
+    }
+
+    /// Decodes a 64 bit float number from `buf` in little endian,
+    /// which is previously encoded via `encode_f64_le`.
+    ///
+    /// # Panics
+    ///
+    /// Panics when `buf.len() < 8`.
+    #[inline]
+    pub fn decode_f64_le(buf: &[u8]) -> f64 {
+        LittleEndian::read_f64(buf)
+    }
+
+    /// Encodes an unsigned 64 bit integer `v` to `buf` in VarInt encoding,
+    /// which is not memory-comparable. Returns thenumber of bytes that encoded.
+    ///
+    /// Note: VarInt encoding is slow, try avoid using it.
+    ///
+    /// # Panics
+    ///
+    /// Panics when `buf.len() < 10`.
+    pub fn encode_var_u64(buf: &mut [u8], mut v: u64) -> usize {
+        assert!(buf.len() >= MAX_VARINT64_LENGTH);
+
+        unsafe {
+            let mut ptr = buf.as_mut_ptr();
+
+            while v >= 0x80 {
+                // We have already checked buffer size at the beginning so that it is safe to
+                // directly access the buffer.
+                *ptr = 0x80 | (v & 0x7f) as u8;
+                ptr = ptr.offset(1);
+                v >>= 7;
+            }
+            *ptr = v as u8;
+            ptr.offset_from(buf.as_ptr()) as usize + 1
+        }
+    }
+
+    /// Decodes an unsigned 64 bit integer from `buf` in VarInt encoding.
+    /// Returns decoded result and the number of bytes that successfully decoded.
+    ///
+    /// This function is more efficient when `buf.len() >= 10`.
+    ///
+    /// # Errors
+    ///
+    /// Returns `Error::BufferTooSmall` if there is not enough space to decode the whole VarInt.
+    pub fn try_decode_var_u64(buf: &[u8]) -> Result<(u64, usize)> {
+        // This efficient implementation is ported from facebook/folly.
+        // Copyright 2013-present Facebook, Inc.
+
+        #[cfg_attr(feature = "cargo-clippy", allow(cast_lossless))]
+        unsafe {
+            let mut val = 0u64;
+            let mut ptr = buf.as_ptr();
+            let ptr_end = buf.as_ptr().offset(buf.len() as isize);
+
+            if ::std::intrinsics::likely(buf.len() >= MAX_VARINT64_LENGTH) {
+                // Fast path
+                let mut b: u64;
+                // Actually this loop never loops, it is only for using `break`.
+                #[cfg_attr(feature = "cargo-clippy", allow(never_loop))]
+                loop {
+                    b = *ptr as u64;
+                    ptr = ptr.offset(1);
+                    val = b & 0x7f;
+                    if b < 0x80 {
```

Comment:
I think the modern compiler is smart enough to expand a loop with small const steps. That is you can use
`for _ in 0..MAX_VARINT64_LENGTH` just like what has been done in tikv version.

Reply:
Tried just now, and seems that the result is not as well as a hand written one.